bpfd/src/lib.rs

```diff
@@ -59,12 +76,13 @@ pub async fn serve(config: Config, static_programs: Vec<StaticPrograms>) -> anyh
     let serve = Server::builder()
         .tls_config(tls_config)?
         .add_service(LoaderServer::with_interceptor(loader, intercept))
-        .serve(addr);
+        .serve_with_shutdown(addr, shutdown_rx.map(drop));
```

Comment:
```suggestion
        .serve_with_shutdown(addr, shutdown_rx.has_changed().map(|_|()));
```

Reply:
This is giving me an error that I'm not smart enough to solve, `serve_with_shutdown` [expects a future](https://docs.rs/tonic/latest/tonic/transport/server/struct.Router.html#method.serve_with_shutdown) and starts shutdown when the future is solved, with this change I get:

```
error[E0277]: `Result<(), tokio::sync::watch::error::RecvError>` is not a future
   --> bpfd/src/lib.rs:79:36
    |
79  |         .serve_with_shutdown(addr, shutdown_rx.has_changed().map(|_| ()));
    |          -------------------       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Result<(), tokio::sync::watch::error::RecvError>` is not a future
    |          |
    |          required by a bound introduced by this call
    |
    = help: the trait `std::future::Future` is not implemented for `Result<(), tokio::sync::watch::error::RecvError>`
    = note: Result<(), tokio::sync::watch::error::RecvError> must be a future or must implement `IntoFuture` to be awaited
note: required by a bound in `Router::<L>::serve_with_shutdown`
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.8.3/src/transport/server/mod.rs:605:41
    |
605 |     pub async fn serve_with_shutdown<F: Future<Output = ()>, ResBody>(
    |                                         ^^^^^^^^^^^^^^^^^^^ required by this bound in `Router::<L>::serve_with_shutdown`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `bpfd` due to previous error
```

All examples on the tonic repo use either `futures` or `futures-util`, I'll look into it a bit further but any help would be appreciated 🙂

Reply:
:( Looks like we need futures for FutureExt, which allows this to compile since you can `.map` on a future.